src/registry.js

```diff
@@ -0,0 +1,133 @@
+import utils from './util';
+const metrics = require('metrics');
+const Report = metrics.Report;
+
+/**
+ * Tagged Metrics Registry.
+ */
+export default class TaggedRegistry extends Report {
+  constructor() {
+    super();
+  }
+
+  /**
+   * Return metric if key exist, otherwise create a new one
+   * @param metricName
+   * @param tags
+   * @param construct
+   * @param supplier
+   * @returns {any | metrics.Meter | metrics.Timer | metrics.Counter | metrics.Histogram | metrics.Gauge|metrics.Metric}
+   */
+  _getOrAddMetric(metricName, tags, construct, supplier = null) {
+    let encodedName = utils.encodeKey(metricName, tags);
+    let metric = this.getMetric(encodedName);
+    if (metric) return metric;
+
+    metric = supplier ? new construct(supplier) : new construct();
```

Comment:
```suggestion
    metric = new construct();
```
Can be simplified.

Reply:
the `supplier` needs to be passed in when the `construct` is a gauge function

Reply:
Oh sorry - I meant this:
```suggestion
    metric = new construct(supplier);
```